In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from sklearn.model_selection import train_test_split
import pickle

In [2]:
raw_df = pd.read_csv('./venom.tsv', sep='\t')

In [32]:
raw_df['Protein families'].head()

0                             Conotoxin I2 superfamily
1                                                  NaN
2    Venom metalloproteinase (M12B) family, P-II su...
3    Venom metalloproteinase (M12B) family, P-II su...
4    Venom metalloproteinase (M12B) family, P-II su...
Name: Protein families, dtype: object

In [33]:
len(raw_df['Protein families'].unique())

608

In [3]:
families_raw = list(raw_df['Protein families'].unique())
families_raw

['Conotoxin I2 superfamily',
 nan,
 'Venom metalloproteinase (M12B) family, P-II subfamily, P-IIa sub-subfamily',
 'Venom metalloproteinase (M12B) family, P-II subfamily, P-IIe sub-subfamily',
 'Venom metalloproteinase (M12B) family, P-II subfamily',
 'Venom metalloproteinase (M12B) family, P-III subfamily',
 'Venom metalloproteinase (M12B) family, P-III subfamily, P-IIIb sub-subfamily',
 'Venom metalloproteinase (M12B) family, P-II subfamily, P-IId sub-subfamily',
 'Venom metalloproteinase (M12B) family, P-III subfamily, P-IIIa sub-subfamily',
 'Venom metalloproteinase (M12B) family, P-III subfamily, P-IIIc sub-subfamily',
 'Venom metalloproteinase (M12B) family, P-II subfamily, P-IIb sub-subfamily',
 'Venom metalloproteinase (M12B) family, P-II subfamily, P-IIc sub-subfamily',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family, P-III subfamily, P-IIId sub-subfamily',
 'Venom metalloproteinase (M12B) family, P-I subfamily',
 'Limacoditoxin-1 (ACP-like) f

In [4]:
families_raw[1] = 'No'

In [5]:
families_list = []
# Truncates protein family names to the family level, discarding subfamily information for simplification and grouping.
for record in families_raw:
        family = record.split(' family')[0] + ' family'
        families_list.append(family)

families_list

['Conotoxin I2 superfamily family',
 'No family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Venom metalloproteinase (M12B) family',
 'Limacoditoxin-1 (ACP-like) family',
 'Xibalbin-2 family',
 'Xibalbin-13 family',
 'Xibalbin-1 family',
 'Three-finger toxin family',
 'Three-finger toxin family',
 'Three-finger toxin family',
 'Three-finger toxin family',
 'Three-finger toxin family',
 'Bradykinin-related peptide family',
 'Short scorpion toxin superfamily, Potassium channel inhibitor kappa-KTx family',
 'Venom waprin family',
 'Ve

In [6]:
families_set = set(families_list)
len(families_set)

308

In [38]:
families_set

{"5'-nucleotidase family",
 'AB hydrolase superfamily, Lipase family',
 'ATP:guanido phosphotransferase family',
 'AVIT (prokineticin) family',
 'Acrorhagin I family',
 'Actinoporin family',
 'Ant myrmeciitoxin-01 family',
 'Ant venom allergen 2/4 family',
 'Arthropod CHH/MIH/GIH/VIH hormone family',
 'Arthropod phospholipase D family',
 'Asilidin-1 family',
 'Asilidin-12 family',
 'Beta-defensin family',
 'Bradykinin inhibitor peptide family',
 'Bradykinin-potentiating peptide family',
 'Bradykinin-related peptide family',
 'CART family',
 'CREC family',
 'CRISP family',
 'Calmodulin family',
 'Caterpillar 1 family',
 'Caterpillar 11 family',
 'Caterpillar 2 family',
 'Caterpillar 3 family',
 'Caterpillar 4 family',
 'Caterpillar 8 family',
 'Caterpillar 9 family',
 'Cathelicidin family',
 'Cationic peptide 01 (latrotoxin) family',
 'Cationic peptide 02 (oxyopinin-2) family',
 'Cationic peptide 03 (latarcin) family',
 'Cationic peptide 04 (cupiennin) family',
 'Cationic peptide 06 (cy

In [39]:
test_df = raw_df.sample(20).copy()
test_df

Entry  Reviewed   Entry Name  \
3154  P51528  reviewed    PA1_VESMC   
4548  Q7Z096  reviewed   I1BC_CONRA   
5068  C0JAX3  reviewed  A1MA1_LOXDE   
1910  C1IC48  reviewed   3S14_WALAE   
3556  D9U297  reviewed   SNA1_LYCMC   
3411  P0C831  reviewed   O166_CONST   
6778  C0HLR6  reviewed   SCX1_RHOCR   
3885  P0DV95  reviewed    TX1_HETPN   
6009  Q9BH86  reviewed   CT14_CONPE   
4672  D2Y2N9  reviewed  H18E1_CYRHA   
1299  P0DQE2  reviewed  TXG1A_CORWE   
1742  A7L035  reviewed  JTX11_CHIFL   
2213  A8YPR6  reviewed   SVMI_ECHOC   
2722  C7C1L2  reviewed   NDB2_OPICY   
5866  P69759  reviewed   O16B_CONVC   
3178  P0DY17  reviewed  NDB43_SCOPA   
960   B6DD23  reviewed  TXD05_LYCSI   
6192  D6C4L9  reviewed   CU92_CONCL   
5180  M9PQ91  reviewed   O164_CONBE   
6229  D6C4H6  reviewed   CU12_CONCL   

                                          Protein names  \
3154  Phospholipase A1 (PLA1) (EC 3.1.1.32) (Allerge...   
4548                   Iota-conotoxin-like r11c (R11.4)   
5068  Dermonecrotic toxin LdSicTox-alphaIB3ai (EC 4....   
1910                            Three-finger toxin W-IV   
3556                                 Neurotoxin LmNaTx1   
3411                          Omega-conotoxin-like S6.6   
6778                                     Beta-toxin Rc1   
3885                      Mu-Sparatoxin-Hp1 (HptTx-208)   
6009                       Conotoxin Pn-014 (Pn-B01131)   
4672                 Hainantoxin-XVIII-5 (HNTX-XVIII-5)   
1299          U-scoloptoxin(16)-Cw1a (U-SLPTX(16)-Cw1a)   
1742                             Toxin CfTX-1 (Toxin 1)   
2213  Snake venom metalloprotease inhibitor 02D01 (0...   
2722               Probable antimicrobial peptide Con13   
5866                             Conotoxin Vc6b (Vc6.2)   
3178                                      Peptide Smp13   
960   U13-lycotoxin-Ls1d (Toxin-like structure LSTX-L5)   
6192                                    Conotoxin Cl9.2   
5180                         Delta-conotoxin-like Bt6.4   
6229                                    Conotoxin Cl1.2   

                                               Organism  Length  \
3154  Vespula maculifrons (Eastern yellow jacket) (W...     300   
4548                        Conus radiatus (Rayed cone)      80   
5068         Loxosceles deserta (Desert recluse spider)     273   
1910        Walterinnesia aegyptia (Desert black snake)      83   
3556      Lychas mucronatus (Chinese swimming scorpion)      86   
3411                     Conus striatus (Striated cone)      72   
6778                  Rhopalurus crassicauda (Scorpion)      49   
3885  Heteropoda pingtungensis (Pingtung huntsman sp...      84   
6009  Conus pennaceus (Feathered cone) (Conus episco...      62   
4672  Cyriopagopus hainanus (Chinese bird spider) (H...     109   
1299  Cormocephalus westwoodi (Westwood's green cent...     109   
1742       Chironex fleckeri (Australian box jellyfish)     456   
2213       Echis ocellatus (Ocellated saw-scaled viper)     308   
2722  Opisthacanthus cayaporum (South American scorp...      81   
5866              Conus victoriae (Queen Victoria cone)      77   
3178  Scorpio palmatus (Israeli golden scorpion) (Sc...      68   
960   Lycosa singoriensis (Wolf spider) (Aranea sing...     120   
6192  Californiconus californicus (California cone) ...      73   
5180                       Conus betulinus (Beech cone)      84   
6229  Californiconus californicus (California cone) ...      28   

                                       Protein families  \
3154            AB hydrolase superfamily, Lipase family   
4548                           Conotoxin I1 superfamily   
5068  Arthropod phospholipase D family, Class II sub...   
1910  Three-finger toxin family, Short-chain subfami...   
3556  Long (4 C-C) scorpion toxin superfamily, Sodiu...   
3411                           Conotoxin O1 superfamily   
6778  Long (4 C-C) scorpion toxin superfamily, Sodiu...   
3885                      Neurotoxin 10 (Hwtx-1) family   
6009    

In [7]:
def simplify_family(df, fam_set):
    df['prot_fams_simp'] = df['Protein families'].copy()
    for family in fam_set:
        for index, row in df.iterrows():
            if family in str(row['prot_fams_simp']):
                df.at[index, 'prot_fams_simp'] = family

In [41]:
simplified_df = raw_df.copy()
simplify_family(simplified_df, families_set)

In [42]:
simplified_df.head()

Entry  Reviewed   Entry Name  \
0  P69499  reviewed     I2_CONST   
1  P0DUS0  reviewed   PRCT_TITSE   
2  C0HLL1  reviewed  VM2CD_CRODU   
3  C0L2T8  reviewed  VM2C1_CRODO   
4  C9E1R9  reviewed  VM2V2_CROVV   

                                       Protein names  \
0                                    conotoxin S11.3   
1      [des-Arg1]-proctolin (NEP-inhibiting peptide)   
2  Zinc metalloproteinase/disintegrin-like CdtV1 ...   
3  Zinc metalloproteinase/disintegrin (Metallopro...   
4  Zinc metalloproteinase/disintegrin VMP-II [Cle...   

                                            Organism  Length  \
0                     Conus striatus (Striated cone)      68   
1             Tityus serrulatus (Brazilian scorpion)       4   
2  Crotalus durissus terrificus (South American r...      32   
3  Crotalus durissus collilineatus (Brazilian rat...     478   
4     Crotalus viridis viridis (Prairie rattlesnake)     478   

                                    Protein families  \
0                           Conotoxin I2 superfamily   
1                                                NaN   
2  Venom metalloproteinase (M12B) family, P-II su...   
3  Venom metalloproteinase (M12B) family, P-II su...   
4  Venom metalloproteinase (M12B) family, P-II su...   

                                   Taxonomic lineage  \
0  cellular organisms (no rank), Eukaryota (domai...   
1  cellular organisms (no rank), Eukaryota (domai...   
2  cellular organisms (no rank), Eukaryota (domai...   
3  cellular organisms (no rank), Eukaryota (domai...   
4  cellular organisms (no rank), Eukaryota (domai...   

                          prot_fams_simp  
0               Conotoxin I2 superfamily  
1                                    NaN  
2  Venom metalloproteinase (M12B) family  
3  Venom metalloproteinase (M12B) family  
4  Venom metalloproteinase (M12B) family

In [43]:
pickle.dump(simplified_df, open('./pickles/simplified_data.pkl', 'wb'))

In [8]:
simplified_df = pickle.load(open('./pickles/simplified_data.pkl', 'rb'))
simplified_df.head()

Entry  Reviewed   Entry Name  \
0  P69499  reviewed     I2_CONST   
1  P0DUS0  reviewed   PRCT_TITSE   
2  C0HLL1  reviewed  VM2CD_CRODU   
3  C0L2T8  reviewed  VM2C1_CRODO   
4  C9E1R9  reviewed  VM2V2_CROVV   

                                       Protein names  \
0                                    conotoxin S11.3   
1      [des-Arg1]-proctolin (NEP-inhibiting peptide)   
2  Zinc metalloproteinase/disintegrin-like CdtV1 ...   
3  Zinc metalloproteinase/disintegrin (Metallopro...   
4  Zinc metalloproteinase/disintegrin VMP-II [Cle...   

                                            Organism  Length  \
0                     Conus striatus (Striated cone)      68   
1             Tityus serrulatus (Brazilian scorpion)       4   
2  Crotalus durissus terrificus (South American r...      32   
3  Crotalus durissus collilineatus (Brazilian rat...     478   
4     Crotalus viridis viridis (Prairie rattlesnake)     478   

                                    Protein families  \
0                           Conotoxin I2 superfamily   
1                                                NaN   
2  Venom metalloproteinase (M12B) family, P-II su...   
3  Venom metalloproteinase (M12B) family, P-II su...   
4  Venom metalloproteinase (M12B) family, P-II su...   

                                   Taxonomic lineage  \
0  cellular organisms (no rank), Eukaryota (domai...   
1  cellular organisms (no rank), Eukaryota (domai...   
2  cellular organisms (no rank), Eukaryota (domai...   
3  cellular organisms (no rank), Eukaryota (domai...   
4  cellular organisms (no rank), Eukaryota (domai...   

                          prot_fams_simp  
0               Conotoxin I2 superfamily  
1                                    NaN  
2  Venom metalloproteinase (M12B) family  
3  Venom metalloproteinase (M12B) family  
4  Venom metalloproteinase (M12B) family

In [45]:
simplified_df['prot_fams_simp'].isnull().sum()

np.int64(539)

In [46]:
simplified_df.query("`prot_fams_simp`.isnull() == True").sort_values('Protein names')

Entry  Reviewed   Entry Name  \
7793  P84745  reviewed   VVP2_BOTJR   
7792  P84747  reviewed   VVP1_BOTMO   
7791  P84746  reviewed   VVP1_BOTJR   
7790  P83717  reviewed  VP239_HETSP   
7789  P83718  reviewed   VP24_HETSP   
...      ...       ...          ...   
232   P0C8B1  reviewed  DEFBL_ORNAN   
218   P0DL40  reviewed  VESPI_VESMG   
154   P0DXZ4  reviewed   TX1A_XYLAR   
153   P0DXZ5  reviewed   TX2A_XYLAR   
1     P0DUS0  reviewed   PRCT_TITSE   

                                          Protein names  \
7793                1095.6 Da venom vasodilator peptide   
7792                1370.6 Da venom vasodilator peptide   
7791                1370.6 Da venom vasodilator peptide   
7790                             2.39 kDa venom peptide   
7789                              2.4 kDa venom peptide   
...                                                 ...   
232   Venom-like beta-defensin (DefB-vL) (Defensin-B...   
218                                              Vespin   
154       Xylotoxin(1)-Xa1a (XYTX(1)-Xa1a) (XYTX1-Xa1a)   
153       Xylotoxin(2)-Xa2a (XYTX(2)-Xa2a) (XYTX2-Xa2a)   
1         [des-Arg1]-proctolin (NEP-inhibiting peptide)   

                                               Organism  Length  \
7793                 Bothrops jararacussu (Jararacussu)      11   
7792   Bothrops moojeni (Lance-headed viper) (Caissaca)      12   
7791                 Bothrops jararacussu (Jararacussu)      12   
7790  Heterometrus spinifer (Asia giant forest scorp...      22   
7789  Heterometrus spinifer (Asia giant forest scorp...      22   
...                                                 ...     ...   
232        Ornithorhynchus anatinus (Duckbill platypus)      68   
218                            Vespa magnifica (Hornet)      67   
154               Xylocopa aruana (Great carpenter bee)      52   
153               Xylocopa aruana (Great carpenter bee)      56   
1                Tityus serrulatus (Brazilian scorpion)       4   

     Protein families                                  Taxonomic lineage  \
7793              NaN  cellular organisms (no rank), Eukaryota (domai...   
7792              NaN  cellular organisms (no rank), Eukaryota (domai...   
7791              NaN  cellular organisms (no rank), Eukaryota (domai...   
7790              NaN  cellular organisms (no rank), Eukaryota (domai...   
7789              NaN  cellular organisms (no rank), Eukaryota (domai...   
...               ...                                                ...   
232               NaN  cellular organisms (no rank), Eukaryota (domai...   
218               NaN  cellular organisms (no rank), Eukaryota (domai...   
154               NaN  cellular organisms (no rank), Eukaryota (domai...   
153               NaN  cellular organisms (no rank), Eukaryota (domai...   
1                 NaN  cellular organisms (no rank), Eukaryota (domai...   

     prot_fams_simp  
7793            NaN  
7792            NaN  
7791            NaN  
7790            NaN  
7789            NaN  
...             ...  
232             NaN  
218             NaN  
154             NaN  
153             NaN  
1               NaN  

[539 rows x 9 columns]

In [ ]:
no_fam_prots_list = list(simplified_df.query("`prot_fams_simp`.isnull() == True").sort_values('Protein names')['Protein names'])
no_fam_prots_list

['1095.6 Da venom vasodilator peptide',
 '1370.6 Da venom vasodilator peptide',
 '1370.6 Da venom vasodilator peptide',
 '2.39 kDa venom peptide',
 '2.4 kDa venom peptide',
 '33kDa venom protein',
 'A-conotoxin cyclic Mila (A-CTX-cMila)',
 'Acatoxin 1 (Acatx1)',
 'Actin depolymerising venom protein gelsolin 1',
 'Alkaline phosphatase (EC 3.1.3.1)',
 'Alpha-actitoxin-Ms11a-1 (Alpha-AITX-Ms11a-1) (Alpha-anmTX-Ms11a-1)',
 'Alpha-actitoxin-Ms11a-2 (Alpha-AITX-Ms11a-2) (Alpha-anmTX-Ms11a-2)',
 'Alpha-actitoxin-Ms11a-3 (Alpha-AITX-Ms11a-3) (Alpha-anmTX-Ms11a-3)',
 'Alpha-actitoxin-Ms11a-4 (Alpha-AITX-Ms11a-4) (Alpha-anmTX-Ms11a-4)',
 'Alpha-conotoxin PrXA (Alpha-C-PrXA)',
 'Alpha-pompilidotoxin (Alpha-PMTX)',
 'Ampulexin 1 (Axn1)',
 'Ampulexin 2 (Axn2)',
 'Ampulexin 3 (Axn3)',
 'Anoplin (ANP) (As-183)',
 'Antimicrobial peptide scolopin-1',
 'Antimicrobial peptide scolopin-2',
 'Antioxidant protein SmP90',
 'Apamin',
 'Apamin (APM) (Apamine)',
 'As-peptide 126 (As-126) (As126)',
 'Astacin-lik

In [ ]:
# In a more rigorous analysis, the proteins without scientifically accepted families would be dropped from the dataset.
# However, this dataset is small enough that preserving even some of these records is beneficial.

# Manually assigned artificial protein families based on proteins with similar names to group proteins without defined families.
# These were generated from substrings of protein names that would group proteins from at least 2 species together.
# To avoid drawing connections that may not exist, names were excluded that grouped distantly related organisms,
# such as scorpions and cone snails.
# Also excluded were names that referred to a protein's discovery or origin rather than its function or structure.
# These include names such as "Protein 1," "Protein 2," etc.,
# as well as "Cryptide", which refers to functional proteins that are created by the breakdown of larger proteins,
# proteins with a "U-" prefix, which indicates that the protein's function is unknown,
# and proteins with "Putative" in the name, which indicates that the protein's function is predicted but not confirmed.
generated_fams = [
    'venom vasodilator peptide',
    'Apamin',
    'LALP',
    'Augerpeptide',
    'buthitoxin',
    'Conantokin',
    'Conopeptide',
    'Conotoxin',
    'Cytolysin',
    'Helokinestatin',
    'Icarapin',
    'Mast cell degranulating peptide',
    'Ornithorhynchus venom',
    'Poneratoxin',
    'Teretoxin',
    'Turripeptide',
    'actitoxin',
    'reduvitoxin',
    'myrmicitoxin',
    'theraphotoxin',
    'hexatoxin'
]

# After reviewing the generated protein families, many were found to be potentially tautological in definition.
# Generated family names that end in -toxin or -peptide are named after the group of organism they came from,
# not necessarily similarity between proteins.
# For instance, "Conotoxin" and "Conopeptide" both refer to proteins found in cone snails,
# but do not necessarily indicate similarity between these proteins.
taut_fams = [
    'Augerpeptide',
    'buthitoxin',
    'Conopeptide',
    'Conotoxin',
    'Ornithorhynchus venom',
    'Poneratoxin',
    'Teretoxin',
    'Turripeptide',
    'actitoxin',
    'reduvitoxin',
    'myrmicitoxin',
    'theraphotoxin',
    'hexatoxin'
]

# These generated protein family groupings were found to represent actual similarities between proteins,
# to the best of the author's knowledge.
# These may be dropped during PCA, but will be used for now.
useful_fams = sorted(list(set(generated_fams) - set(taut_fams)))
useful_fams

['Apamin',
 'Conantokin',
 'Cytolysin',
 'Helokinestatin',
 'Icarapin',
 'LALP',
 'Mast cell degranulating peptide',
 'venom vasodilator peptide']

In [9]:
pd.set_option('display.max_rows', 200)

In [151]:
# Used to check potential generated family names for usefulness.
simplified_df[simplified_df['Protein names'].str.contains(
    "LALP" # Put potential name as a string here
    )].query('`prot_fams_simp`.isnull() == True')

Entry  Reviewed   Entry Name  \
7181  A0FKN6  reviewed   VMPA_LOXIN   
7182  C9D7R2  reviewed  VMPA2_LOXIN   
7183  C9D7R3  reviewed  VMPA3_LOXIN   
7184  P0DM61  reviewed  VMPA4_LOXLA   
7185  P0DM62  reviewed  VMPA5_LOXGA   

                                          Protein names  \
7181  Astacin-like metalloprotease toxin 1 (EC 3.4.2...   
7182  Astacin-like metalloprotease toxin 2 (EC 3.4.2...   
7183  Astacin-like metalloprotease toxin 3 (EC 3.4.2...   
7184  Astacin-like metalloprotease toxin 4 (EC 3.4.2...   
7185  Astacin-like metalloprotease toxin 5 (EC 3.4.2...   

                                               Organism  Length  \
7181               Loxosceles intermedia (Brown spider)     264   
7182               Loxosceles intermedia (Brown spider)     256   
7183               Loxosceles intermedia (Brown spider)     243   
7184  Loxosceles laeta (South American recluse spide...     114   
7185                         Loxosceles gaucho (Spider)     186   

     Protein families                                  Taxonomic lineage  \
7181              NaN  cellular organisms (no rank), Eukaryota (domai...   
7182              NaN  cellular organisms (no rank), Eukaryota (domai...   
7183              NaN  cellular organisms (no rank), Eukaryota (domai...   
7184              NaN  cellular organisms (no rank), Eukaryota (domai...   
7185              NaN  cellular organisms (no rank), Eukaryota (domai...   

     prot_fams_simp  
7181            NaN  
7182            NaN  
7183            NaN  
7184            NaN  
7185            NaN

In [152]:
def simplify_null_family(df, fam_set):
    for family in fam_set:
        for index, row in df.iterrows():
            if family in str(row['Protein names']) and pd.isnull(row['prot_fams_simp']):
                df.at[index, 'prot_fams_simp'] = family

In [153]:
test_df = simplified_df.copy()
test_df

Entry  Reviewed   Entry Name  \
0     P69499  reviewed     I2_CONST   
1     P0DUS0  reviewed   PRCT_TITSE   
2     C0HLL1  reviewed  VM2CD_CRODU   
3     C0L2T8  reviewed  VM2C1_CRODO   
4     C9E1R9  reviewed  VM2V2_CROVV   
...      ...       ...          ...   
7789  P83718  reviewed   VP24_HETSP   
7790  P83717  reviewed  VP239_HETSP   
7791  P84746  reviewed   VVP1_BOTJR   
7792  P84747  reviewed   VVP1_BOTMO   
7793  P84745  reviewed   VVP2_BOTJR   

                                          Protein names  \
0                                       conotoxin S11.3   
1         [des-Arg1]-proctolin (NEP-inhibiting peptide)   
2     Zinc metalloproteinase/disintegrin-like CdtV1 ...   
3     Zinc metalloproteinase/disintegrin (Metallopro...   
4     Zinc metalloproteinase/disintegrin VMP-II [Cle...   
...                                                 ...   
7789                              2.4 kDa venom peptide   
7790                             2.39 kDa venom peptide   
7791                1370.6 Da venom vasodilator peptide   
7792                1370.6 Da venom vasodilator peptide   
7793                1095.6 Da venom vasodilator peptide   

                                               Organism  Length  \
0                        Conus striatus (Striated cone)      68   
1                Tityus serrulatus (Brazilian scorpion)       4   
2     Crotalus durissus terrificus (South American r...      32   
3     Crotalus durissus collilineatus (Brazilian rat...     478   
4        Crotalus viridis viridis (Prairie rattlesnake)     478   
...                                                 ...     ...   
7789  Heterometrus spinifer (Asia giant forest scorp...      22   
7790  Heterometrus spinifer (Asia giant forest scorp...      22   
7791                 Bothrops jararacussu (Jararacussu)      12   
7792   Bothrops moojeni (Lance-headed viper) (Caissaca)      12   
7793                 Bothrops jararacussu (Jararacussu)      11   

                                       Protein families  \
0                              Conotoxin I2 superfamily   
1                                                   NaN   
2     Venom metalloproteinase (M12B) family, P-II su...   
3     Venom metalloproteinase (M12B) family, P-II su...   
4     Venom metalloproteinase (M12B) family, P-II su...   
...                                                 ...   
7789                                                NaN   
7790                                                NaN   
7791                                                NaN   
7792                                                NaN   
7793                                                NaN   

                                      Taxonomic lineage  \
0     cellular organisms (no rank), Eukaryota (domai...   
1     cellular organisms (no rank), Eukaryota (domai...   
2     cellular organisms (no rank), Eukaryota (domai...   
3     cellular organisms (no rank), Eukaryota (domai...   
4     cellular organisms (no rank), Eukaryota (domai...   
...                                                 ...   
7789  cellular organisms (no rank), Eukaryota (domai...   
7790  cellular organisms (no rank), Eukaryota (domai...   
7791  cellular organisms (no rank), Eukaryota (domai...   
7792  cellular organisms (no rank), Eukaryota (domai...   
7793  cellular organisms (no rank), Eukaryota (domai...   

                             prot_fams_simp  
0                  Conotoxin I2 superfamily  
1                                       NaN  
2     Venom metalloproteinase (M12B) family  
3     Venom metalloproteinase (M12B) family  
4     Venom metalloproteinase (M12B) family  
...                                     ...  
7789                                    NaN  
7790                                    NaN  
7791                                    NaN  
7792                                    NaN  
7793                                    NaN  

[7794 rows x 9 columns]

In [187]:
simplify_null_family(simplified_df, useful_fams)

In [188]:
simplified_df.query("`prot_fams_simp` == 'LALP'")

Entry  Reviewed   Entry Name  \
7181  A0FKN6  reviewed   VMPA_LOXIN   
7182  C9D7R2  reviewed  VMPA2_LOXIN   
7183  C9D7R3  reviewed  VMPA3_LOXIN   
7184  P0DM61  reviewed  VMPA4_LOXLA   
7185  P0DM62  reviewed  VMPA5_LOXGA   

                                          Protein names  \
7181  Astacin-like metalloprotease toxin 1 (EC 3.4.2...   
7182  Astacin-like metalloprotease toxin 2 (EC 3.4.2...   
7183  Astacin-like metalloprotease toxin 3 (EC 3.4.2...   
7184  Astacin-like metalloprotease toxin 4 (EC 3.4.2...   
7185  Astacin-like metalloprotease toxin 5 (EC 3.4.2...   

                                               Organism  Length  \
7181               Loxosceles intermedia (Brown spider)     264   
7182               Loxosceles intermedia (Brown spider)     256   
7183               Loxosceles intermedia (Brown spider)     243   
7184  Loxosceles laeta (South American recluse spide...     114   
7185                         Loxosceles gaucho (Spider)     186   

     Protein families                                  Taxonomic lineage  \
7181              NaN  cellular organisms (no rank), Eukaryota (domai...   
7182              NaN  cellular organisms (no rank), Eukaryota (domai...   
7183              NaN  cellular organisms (no rank), Eukaryota (domai...   
7184              NaN  cellular organisms (no rank), Eukaryota (domai...   
7185              NaN  cellular organisms (no rank), Eukaryota (domai...   

     prot_fams_simp  
7181           LALP  
7182           LALP  
7183           LALP  
7184           LALP  
7185           LALP

In [184]:
test_df = simplified_df[['Organism', 'prot_fams_simp', 'Taxonomic lineage']].head(400).sort_values('Organism').copy()
test_df['prot_pres'] = 1
test_df = test_df.dropna()
test_df = test_df.drop_duplicates()
test_df

Organism  \
300  Agelenopsis aperta (North American funnel-web ...   
85   Agkistrodon bilineatus (Cantil) (Tropical mocc...   
347  Agkistrodon contortrix contortrix (Southern co...   
42   Agkistrodon contortrix contortrix (Southern co...   
82   Agkistrodon contortrix laticinctus (Broad-band...   
66   Agkistrodon piscivorus leucostoma (Western cot...   
293  Agkistrodon piscivorus leucostoma (Western cot...   
33   Agkistrodon piscivorus piscivorus (Eastern cot...   
234                      Apis cerana (Indian honeybee)   
238                          Apis mellifera (Honeybee)   
193                          Apis mellifera (Honeybee)   
239                          Apis mellifera (Honeybee)   
377                        Azemiops feae (Fea's viper)   
103                Bitis arietans (African puff adder)   
29        Bitis gabonica (Gaboon adder) (Gaboon viper)   
237                         Bombus ignitus (Bumblebee)   
235                         Bombus ignitus (Bumblebee)   
287  Bombus pensylvanicus (American bumblebee) (Api...   
286  Bombus terrestris (Buff-tailed bumblebee) (Api...   
150  Borikenophis portoricensis (Puerto Rican racer...   
139  Bothriechis lateralis (Side-striped palm pitvi...   
88   Bothrops alternatus (Urutu) (Rhinocerophis alt...   
23                         Bothrops asper (Terciopelo)   
59      Bothrops atrox (Barba amarilla) (Fer-de-lance)   
148  Bothrops barnetti (Barnett's lancehead) (Trime...   
361  Bothrops cotiara (Cotiara) (Rhinocerophis coti...   
129        Bothrops erythromelas (Caatinga lance head)   
27   Bothrops insularis (Golden lancehead) (Lachesi...   
112   Bothrops jararaca (Jararaca) (Bothrops jajaraca)   
115                 Bothrops jararacussu (Jararacussu)   
359                 Bothrops jararacussu (Jararacussu)   
145            Bothrops leucurus (Whitetail lancehead)   
97    Bothrops moojeni (Lance-headed viper) (Caissaca)   
174                  Bungarus candidus (Malayan krait)   
53   Bungarus fasciatus (Banded krait) (Pseudoboa f...   
353          Bungarus multicinctus (Many-banded krait)   
54           Bungarus multicinctus (Many-banded krait)   
20   Calloselasma rhodostoma (Malayan pit viper) (A...   
71           Cerberus rynchops (Dog-faced water snake)   
322                   Chaerilus tricostatus (Scorpion)   
0                       Conus striatus (Striated cone)   
17   Craspedocephalus gramineus (Bamboo pit viper) ...   
80   Crotalus adamanteus (Eastern diamondback rattl...   
216  Crotalus adamanteus (Eastern diamondback rattl...   
297  Crotalus adamanteus (Eastern diamondback rattl...   
130   Crotalus atrox (Western diamondback rattlesnake)   
58   Crotalus durissus cascavella (Northeastern Bra...   
296  Crotalus durissus collilineatus (Brazilian rat...   
3    Crotalus durissus collilineatus (Brazilian rat...   
118  Crotalus durissus durissus (Central American r...   
360  Crotalus durissus terrificus (South American r...   
109  Crotalus durissus terrificus (South American r...   
48   Crotalus scutulatus scutulatus (Mojave rattles...   
57   Crotalus vegrandis (Uracoan rattlesnake) (Crot...   
89      Crotalus viridis viridis (Prairie rattlesnake)   
356  Cryptophis nigrescens (Eastern small-eyed snak...   
256  Cryptophis nigrescens (Eastern small-eyed snak...   
208  Daboia russelii (Russel's viper) (Vipera russe...   
56   Daboia russelii (Russel's viper) (Vipera russe...   
352  Daboia russelii (Russel's viper) (Vipera russe...   
96   Daboia siamensis (Eastern Russel's viper) (Dab...   
45   Deinagkistrodon acutus (Hundred-pace snake) (A...   
364  Demansia vestigiata (Lesser black whip snake) ...   
210  Demansia vestigiata (Lesser black whip snake) ...   
242  Demansia vestigiata (Lesser black whip snake) ...   
263    Dendroaspis jamesoni jamesoni (Jameson's mamba)   
151            Doratifera vulnerans (Mottled cup moth)   
72   Drysdalia coronoides (White-lipped snake) (Hop...   
348  Drysdalia coronoides (White-lipped snak

In [185]:
test_df = test_df.pivot(index='Organism', columns='prot_fams_simp', values='prot_pres')
test_df = test_df.fillna(0)
test_df

prot_fams_simp                                      AVIT (prokineticin) family  \
Organism                                                                         
Agelenopsis aperta (North American funnel-web s...                         0.0   
Agkistrodon bilineatus (Cantil) (Tropical mocca...                         0.0   
Agkistrodon contortrix contortrix (Southern cop...                         0.0   
Agkistrodon contortrix laticinctus (Broad-bande...                         0.0   
Agkistrodon piscivorus leucostoma (Western cott...                         0.0   
Agkistrodon piscivorus piscivorus (Eastern cott...                         0.0   
Apis cerana (Indian honeybee)                                              0.0   
Apis mellifera (Honeybee)                                                  0.0   
Azemiops feae (Fea's viper)                                                0.0   
Bitis arietans (African puff adder)                                        0.0   
Bitis gabonica (Gaboon adder) (Gaboon viper)                               0.0   
Bombus ignitus (Bumblebee)                                                 0.0   
Bombus pensylvanicus (American bumblebee) (Apis...                         0.0   
Bombus terrestris (Buff-tailed bumblebee) (Apis...                         0.0   
Borikenophis portoricensis (Puerto Rican racer)...                         0.0   
Bothriechis lateralis (Side-striped palm pitviper)                         0.0   
Bothrops alternatus (Urutu) (Rhinocerophis alte...                         0.0   
Bothrops asper (Terciopelo)                                                0.0   
Bothrops atrox (Barba amarilla) (Fer-de-lance)                             0.0   
Bothrops barnetti (Barnett's lancehead) (Trimer...                         0.0   
Bothrops cotiara (Cotiara) (Rhinocerophis cotiara)                         0.0   
Bothrops erythromelas (Caatinga lance head)                                0.0   
Bothrops insularis (Golden lancehead) (Lachesis...                         0.0   
Bothrops jararaca (Jararaca) (Bothrops jajaraca)                           0.0   
Bothrops jararacussu (Jararacussu)                                         0.0   
Bothrops leucurus (Whitetail lancehead)                                    0.0   
Bothrops moojeni (Lance-headed viper) (Caissaca)                           0.0   
Bungarus candidus (Malayan krait)                                          0.0   
Bungarus fasciatus (Banded krait) (Pseudoboa fa...                         0.0   
Bungarus multicinctus (Many-banded krait)                                  0.0   
Calloselasma rhodostoma (Malayan pit viper) (Ag...                         0.0   
Cerberus rynchops (Dog-faced water snake)                                  0.0   
Chaerilus tricostatus (Scorpion)                                           0.0   
Conus striatus (Striated cone)                                             0.0   
Craspedocephalus gramineus (Bamboo pit viper) (...                         0.0   
Crotalus adamanteus (Eastern diamondback rattle...                         0.0   
Crotalus atrox (Western diamondback rattlesnake)                           0.0   
Crotalus durissus cascavella (Northeastern Braz...                         0.0   
Crotalus durissus collilineatus (Brazilian ratt...                         0.0   
Crotalus durissus durissus (Central American ra...                         0.0   
Crotalus durissus terrificus (South American ra...                         0.0   
Crotalus scutulatus scutulatus (Mojave rattlesn...                         0.0   
Crotalus vegrandis (Uracoan rattlesnake) (Crota...                         0.0   
Crotalus viridis viridis (Prairie rattlesnake)                             0.0   
Cryptophis nigrescens (Eastern small-eyed snake...                         0.0   
Daboia russelii (Russel's viper) (Vipera russelii)                         0.0   
Daboia siamensis (Eastern Russel's viper) (Dabo...              

In [192]:
organism_pivoted_df = simplified_df[['Organism', 'prot_fams_simp', 'Taxonomic lineage']].sort_values('Organism').copy()
organism_pivoted_df['prot_pres'] = 1
print(organism_pivoted_df.shape)
organism_pivoted_df = organism_pivoted_df.dropna()
print(organism_pivoted_df.shape)
organism_pivoted_df = organism_pivoted_df.drop_duplicates()
print(organism_pivoted_df.shape)

organism_pivoted_df = organism_pivoted_df.pivot(index='Organism', columns='prot_fams_simp', values='prot_pres')
organism_pivoted_df = organism_pivoted_df.fillna(0)
print(organism_pivoted_df.shape)
organism_pivoted_df.head()

(7794, 4)
(7279, 4)
(2421, 4)
(830, 315)


prot_fams_simp                                      5'-nucleotidase family  \
Organism                                                                     
Acanthaster planci (Crown-of-thorns starfish)                          0.0   
Acanthophis antarcticus (Common death adder)                           0.0   
Acanthophis rugosus (Irian Jayan death adder)                          0.0   
Acanthophis sp. (strain Seram) (Seram death adder)                     0.0   
Acanthoscurria gomesiana (Tarantula spider) (Ph...                     0.0   

prot_fams_simp                                      AB hydrolase superfamily, Lipase family  \
Organism                                                                                      
Acanthaster planci (Crown-of-thorns starfish)                                           0.0   
Acanthophis antarcticus (Common death adder)                                            0.0   
Acanthophis rugosus (Irian Jayan death adder)                                           0.0   
Acanthophis sp. (strain Seram) (Seram death adder)                                      0.0   
Acanthoscurria gomesiana (Tarantula spider) (Ph...                                      0.0   

prot_fams_simp                                      ATP:guanido phosphotransferase family  \
Organism                                                                                    
Acanthaster planci (Crown-of-thorns starfish)                                         0.0   
Acanthophis antarcticus (Common death adder)                                          0.0   
Acanthophis rugosus (Irian Jayan death adder)                                         0.0   
Acanthophis sp. (strain Seram) (Seram death adder)                                    0.0   
Acanthoscurria gomesiana (Tarantula spider) (Ph...                                    0.0   

prot_fams_simp                                      AVIT (prokineticin) family  \
Organism                                                                         
Acanthaster planci (Crown-of-thorns starfish)                              0.0   
Acanthophis antarcticus (Common death adder)                               0.0   
Acanthophis rugosus (Irian Jayan death adder)                              0.0   
Acanthophis sp. (strain Seram) (Seram death adder)                         0.0   
Acanthoscurria gomesiana (Tarantula spider) (Ph...                         0.0   

prot_fams_simp                                      Acrorhagin I family  \
Organism                                                                  
Acanthaster planci (Crown-of-thorns starfish)                       0.0   
Acanthophis antarcticus (Common death adder)                        0.0   
Acanthophis rugosus (Irian Jayan death adder)                       0.0   
Acanthophis sp. (strain Seram) (Seram death adder)                  0.0   
Acanthoscurria gomesiana (Tarantula spider) (Ph...                  0.0   

prot_fams_simp                                      Actinoporin family  \
Organism                                                                 
Acanthaster planci (Crown-of-thorns starfish)                      0.0   
Acanthophis antarcticus (Common death adder)                       0.0   
Acanthophis rugosus (Irian Jayan death adder)                      0.0   
Acanthophis sp. (strain Seram) (Seram death adder)                 0.0   
Acanthoscurria gomesiana (Tarantula spider) (Ph...                 0.0   

prot_fams_simp                                      Ant myrmeciitoxin-01 family  \
Organism                                                                          
Acanthaster planci (Crown-of-thorns starfish)                               0.0   
Acanthophis antarcticus (Common death adder)                                0.0   
Acanthophis rugosus (Irian Jayan death adder)                               0.0   
Acanthophis sp. (strain Seram) (Seram death adder)                          0.0   
Acanthoscurria go

In transforming the proteins dataset to a form where each species is a single record instead of each protein being a single record, a couple of steps were executed, each of which reduced the size of the dataset.
- Group proteins by family and drop proteins that do not fit into a family.
    - Dropping null family proteins reduced the number of records from 7794 to 7279.
    - Grouping the proteins by family and removing duplicates (within each species) reduced the number of records from 7279 to 2421.
- Pivoting the dataset so that each species was a single record and each protein family became its own column transformed the dataset from 2421 records and 6 columns to 830 records (species) and 315 columns (protein families)

In [206]:
print(organism_pivoted_df.sum().mean())
print(organism_pivoted_df.sum().mode())
print(organism_pivoted_df.sum().median())
organism_pivoted_df.describe()

7.685714285714286
0    1.0
dtype: float64
2.0


prot_fams_simp  5'-nucleotidase family  \
count                       830.000000   
mean                          0.006024   
std                           0.077428   
min                           0.000000   
25%                           0.000000   
50%                           0.000000   
75%                           0.000000   
max                           1.000000   

prot_fams_simp  AB hydrolase superfamily, Lipase family  \
count                                        830.000000   
mean                                           0.021687   
std                                            0.145746   
min                                            0.000000   
25%                                            0.000000   
50%                                            0.000000   
75%                                            0.000000   
max                                            1.000000   

prot_fams_simp  ATP:guanido phosphotransferase family  \
count                                      830.000000   
mean                                         0.001205   
std                                          0.034711   
min                                          0.000000   
25%                                          0.000000   
50%                                          0.000000   
75%                                          0.000000   
max                                          1.000000   

prot_fams_simp  AVIT (prokineticin) family  Acrorhagin I family  \
count                           830.000000           830.000000   
mean                              0.004819             0.001205   
std                               0.069295             0.034711   
min                               0.000000             0.000000   
25%                               0.000000             0.000000   
50%                               0.000000             0.000000   
75%                               0.000000             0.000000   
max                               1.000000             1.000000   

prot_fams_simp  Actinoporin family  Ant myrmeciitoxin-01 family  \
count                   830.000000                   830.000000   
mean                      0.025301                     0.001205   
std                       0.157133                     0.034711   
min                       0.000000                     0.000000   
25%                       0.000000                     0.000000   
50%                       0.000000                     0.000000   
75%                       0.000000                     0.000000   
max                       1.000000                     1.000000   

prot_fams_simp  Ant venom allergen 2/4 family      Apamin  \
count                              830.000000  830.000000   
mean                                 0.004819    0.002410   
std                                  0.069295    0.049058   
min                                  0.000000    0.000000   
25%                                  0.000000    0.000000   
50%                                  0.000000    0.000000   
75%                                  0.000000    0.000000   
max                                  1.000000    1.000000   

prot_fams_simp  Arthropod CHH/MIH/GIH/VIH hormone family  ...  \
count                                         830.000000  ...   
mean                                            0.004819  ...   
std                                             0.069295  ...   
min                                             0.000000  ...   
25%                                             0.000000  ...   
50%                                             0.000000  ...   
75%                                             0.000000  ...   
max                                             1.000000  ...   

prot_fams_simp  Venom complement C3 homolog family  \
count                                   830.000000   
mean                                      0.004819   
std                                       0.069295   
min            

In [207]:
for col in organism_pivoted_df.columns:
    if organism_pivoted_df[col].sum() < 2:
        print(col, organism_pivoted_df[col].sum())

ATP:guanido phosphotransferase family 1.0
Acrorhagin I family 1.0
Ant myrmeciitoxin-01 family 1.0
Asilidin-12 family 1.0
CART family 1.0
CREC family 1.0
Caterpillar 1 family 1.0
Caterpillar 11 family 1.0
Caterpillar 2 family 1.0
Caterpillar 3 family 1.0
Caterpillar 4 family 1.0
Caterpillar 8 family 1.0
Caterpillar 9 family 1.0
Cationic peptide 02 (oxyopinin-2) family 1.0
Cationic peptide 03 (latarcin) family 1.0
Cationic peptide 06 (cytoinsectotoxin) family 1.0
ConoGAY family 1.0
Conopeptide I2-like superfamily 1.0
Conotoxin B2 family 1.0
Conotoxin E superfamily 1.0
Conotoxin F superfamily 1.0
Conotoxin G superfamily 1.0
Conotoxin G2 superfamily, 1 family 1.0
Conotoxin K superfamily 1.0
Conotoxin M family 1.0
Conotoxin NSf-1 superfamily 1.0
Conotoxin O1 family 1.0
Conotoxin O2 family 1.0
Conotoxin Y superfamily 1.0
DNase II family 1.0
Damage-control phosphatase family 1.0
Dermatopontin family 1.0
Disintegrin-like family 1.0
F superfamily 1.0
Gastrin/cholecystokinin family 1.0
Glycoprot

In [208]:
simplified_df['Taxonomic lineage'].unique()

array(['cellular organisms (no rank), Eukaryota (domain), Opisthokonta (clade), Metazoa (kingdom), Eumetazoa (clade), Bilateria (clade), Protostomia (clade), Spiralia (clade), Lophotrochozoa (clade), Mollusca (phylum), Gastropoda (class), Caenogastropoda (subclass), Neogastropoda (order), Conoidea (superfamily), Conidae (family), Conus (genus), Pionoconus (subgenus)',
       'cellular organisms (no rank), Eukaryota (domain), Opisthokonta (clade), Metazoa (kingdom), Eumetazoa (clade), Bilateria (clade), Protostomia (clade), Ecdysozoa (clade), Panarthropoda (clade), Arthropoda (phylum), Chelicerata (subphylum), Arachnida (class), Scorpiones (order), Buthida (parvorder), Buthoidea (superfamily), Buthidae (family), Tityus (genus)',
       'cellular organisms (no rank), Eukaryota (domain), Opisthokonta (clade), Metazoa (kingdom), Eumetazoa (clade), Bilateria (clade), Deuterostomia (clade), Chordata (phylum), Craniata (subphylum), Vertebrata (clade), Gnathostomata (clade), Teleostomi (clade)